server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -1055,7 +1059,7 @@ private IndexResult indexIntoLucene(Index index, IndexingStrategy plan)
             }
             return new IndexResult(plan.versionForIndexing, index.primaryTerm(), index.seqNo(), plan.currentNotFoundOrDeleted);
         } catch (Exception ex) {
-            if (indexWriter.getTragicException() == null) {
+            if ( treatDocumentFailureAsTragicError(index) == false && indexWriter.getTragicException() == null) {
```

Comment:
should we treat `AlreadyClosedException` specially here as well (same as when we index deletion or noop tombstone).

Reply:
We should not have a special treatment for AlreadyClosedException here. If the engine was failed and closed by other thread, it's perfectly fine to bubble up to the AlreadyClosedException. In fact, we should bubble up AlreadyClosedException so we can [detect](https://github.com/elastic/elasticsearch/blob/4af1649aa8696c0e384fbc4414616368ed51355b/server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java#L1997-L2000) situations where the engine does not have a buggy state.

However, we probably should call `maybeFailEngine` instead of `failEngine` if the exception is AlreadyClosedEngine to avoid unnecessary [warning log](https://github.com/elastic/elasticsearch/blob/399d53e5c03ae4157d9ba107d01fbecfed758972/server/src/main/java/org/elasticsearch/index/engine/Engine.java#L1157-L1162) if the engine was failed already.

Reply:
I think we should not try to wrap AlreadyClosedException into an IndexResult as we might possibly write it to the translog during closing.

Reply:
++. Fixed in https://github.com/elastic/elasticsearch/pull/43523/commits/c6325268009cc909bf1fadd92ca020f60a42efff.